modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheEntryListenersRegistry.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.client.thin;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.cache.configuration.CacheEntryListenerConfiguration;
+
+/**
+ * Per-cache cache entry listeners registry. Listeners can't be stored inside ClientCache instance, since there can be
+ * several such instances per one cache.
+ */
+public class ClientCacheEntryListenersRegistry {
```

Comment:
I think we should do the following:
* Refactor existing `TcpClientChannel.notificationLsnrs` to a `Map<Long, NotificationListener>`
* Instead of iterating over all `notificationLsnrs` and checking the id in every listener, get the listener by id from map and invoke it without inner checks
* Use this mechanism for continuous query notifications, avoid creating a new registry

This approach seems to work fine in .NET thin client.

Reply:
This will require creating new listener for each compute task and refactoring of existing compute functionality (I think if we agreed to do this, it should be done by another ticket). Also, this change will add new limitations: we can't use notifications without resource id in the future.
To reduce count of listeners, I think for continuous queries we can act the same way as for compute: register only one listener for QUERY_CONTINUOUS_EVENT_NOTIFICATION and find handles in this listener by resource id using some map. WDYT?

Reply:
> new limitations: we can't use notifications without resource id in the future

As I understand, this limitation already exists - all messages have an ID in any case.

> find handles in this listener by resource id using some map

We could do this, but again, why have `Collection<NotificationListener>`, then separate maps in Compute and Continuous Query (and, possibly, more places in future), when a single map will:
* Simplify the implementation
* Make it more efficient (single lookup, single collection)

Reply:
>  it should be done by another ticket

Ok, let's do this first, and then get back to the current ticket?

Reply:
> As I understand, this limitation already exists - all messages have an ID in any case.
There is no requirement for the uniqueness of this ID. We can use 0 for some events (for example, we have topology change event implemented as special flag of every response, but it can be implemented with notification and no additional resource ID required here). Also, perhaps in the future events initiated by server-side for some resource will be needed, in this case, client should be able to listen some type of notification with any resource ID. 
> why have Collection<NotificationListener>, then separate maps in Compute and Continuous Query
We can replace the first collection with `EnumMap` (Listener per ClientOperation), it is very lightweight and fast, and there will be no iterations. There will be at most one listener per one operation type, and each listener can implement its own strategy for resource ID mapping. This change does not require to refactor compute a lot, so, I think, it can be implemented in the current ticket. 
Or we can even provide two API on client channel level: listener for client operation and any resource id (refactor compute to use this API) and listener for client operation and exactly one resource Id. WDYT?

Reply:
> There is no requirement for the uniqueness of this ID

Right now this ID is required to match either a compute task ID, or a continuous query ID.
A combination of `resId`+`notificationCode` is required to be "unique". We can use `IgniteBiTuple` as a map key.
If we are going to introduce predefined ids (like 0 or -1) for general-purpose events - that will work fine as well.

> each listener can implement its own strategy for resource ID mapping

This looks like code duplication to me.

Reply:
> If we are going to introduce predefined ids (like 0 or -1) for general-purpose events - that will work fine as well.

There is still a case with resource IDs initiated from the server, but ok, we don't have it now, if we will need it we can change implementation.
I've tried to change notification listeners and listen to resource ID, but unfortunately, there is one critical problem: we can process notification before the notification listener is registered. For compute it's a blocker since notification can be lost and the future will never complete. This happens because the earliest place when we can register listener - it's payload reader, payload reader executed in user's thread or in `asyncContinuationExecutor` thread. Concurrently the new message (notification) from the channel can be read by NIO thread and executed by `asyncContinuationExecutor` without notifying listeners.

Reply:
The same is true for Continuous Query - you can get a notification before you know the query ID and register the listener (and you can have multiple active queries).
This is another piece of logic to be reused - that's why we need a centralized notification handling mechanism.

We can queue "unknown" notifications per resourceId and handle them as soon as listener is registered - we'll have to do it one way or another. For Compute the queue will have at most 1 item, but for CQ it can have any number of events.

Reply:
Please have a look at [C# implementation](https://github.com/apache/ignite/blob/master/modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientNotificationHandler.cs) - hopefully I did not miss anything there.

The only potential problem is that `requestId` is assumed to be unique across Compute, CQ and other future notification types - we should use a combined key instead (notificationType + requestId).

Reply:
> The same is true for Continuous Query - you can get a notification before you know the query ID and register the listener (and you can have multiple active queries).

For CQ we don't need such strong guarantees. We can skip some events from the start and CQ still will be valid (it's like if we start listening a little bit later).

> This is another piece of logic to be reused.
> We can queue "unknown" notifications per resourceId and handle them as soon as listener is registered

The difference between CQ and compute: compute can't skip notifications, compute has exactly one notification, taking into account this logic compute was implemented in such way, that unregistering listener performed when receiving notification. For CQ there can be many notifications per listener and listener unregistering should be performed when we close CQ. And if we will use some queue for pending notifications there will be notifications leaks after CQ closing, some notifications can stay in queue forever (as far as I understand C# implementation has the same problem). We can use some flag for notification type (use pending notification queue for compute and don't use it for CQ) to solve this problem, but I'm not sure that it's better than solution with different logic for different components.

Reply:
@ptupitsyn I've refactored notification listeners and partially move changes to the ticket 
https://issues.apache.org/jira/browse/IGNITE-14492. When this ticket will be in the master branch I will rebase the current ticket.

Reply:
@ptupitsyn, all comments fixed, ticket rebased, can you please have a look again?